status-indicator: Show actual cloning state

mrnugget opened the pull request:
This changes the status indicator from reporting whether the repositories are updating to including their actual cloning status.

This is an experiment, since the cloning status is computed with every request, every 5 seconds, by loading the complete list of repository names from the database and asking the gitserver shards whether they are cloned. Each gitserver instance then has to do a maximum of 2 `os.Stat` calls per repository name.

The big upside is that this reports the actual "is this repository cloned" status. A tiny asterisk: just because a repository is currently _not cloned_ doesn't mean it's cloning. But that could be fixed by request a new update for each uncloned repository we come across.

Questions/TODOs:
- [x] the metrics for the new `ListAllRepoNames` don't show up in `repo-updater`'s `/metrics` endpoint
- [ ] the code in the `gitserver.Client` method `AreReposCloned` is duplicated from `RepoInfo` since it needs to use the same sharding logic. This should probably be refactored
- [x] No tests

tsenart commented:
@mrnugget: Regarding the metrics, you need to register them: https://github.com/sourcegraph/sourcegraph/blob/4b99c384c314b34640c9e8f597055a09d815dd85/cmd/repo-updater/main.go#L70

tsenart commented on pkg/gitserver/client.go:
```diff
@@ -529,6 +529,78 @@ func (c *Client) IsRepoCloned(ctx context.Context, repo api.RepoName) (bool, err
+			o.err = json.NewDecoder(resp.Body).Decode(o.res)
+			ch <- o
+
+			ch <- o
```
Double send. Ooops

mrnugget replied:
Nice.

mrnugget commented:
I ran a few tests on my MacBook to get ballpark figures for how fast/slow it is to ask gitserver about the cloned status for multiple thousands of repositories.

First I created dummy request data that makes `gitserver` check each repository name with the maximum of two `os.Stat` calls since none of the repositories exists on my local disk:

```bash
ruby -e 'require "json"; puts JSON.dump({"Repos" => (0..30_000).map {|i| "github.com/testing/repo-#{i}"}})' | jq . > ~/tmp/30_000_repos_are_cloned.json
```

Then I sent that to the local `gitserver` instance (no sharding!):

```bash
time curl -s 'http://localhost:3178/are-repos-cloned' -X POST -H 'Content-Type: application/json' --data @/Users/thorstenball/tmp/30_000_repos_are_cloned.json >/dev/null
```

Here are the results for 10+ runs:

```
0.01s user 0.01s system 3% cpu 0.462 total
0.01s user 0.01s system 5% cpu 0.329 total
0.01s user 0.01s system 5% cpu 0.329 total
0.01s user 0.01s system 5% cpu 0.327 total
0.01s user 0.01s system 6% cpu 0.340 total
0.01s user 0.01s system 5% cpu 0.354 total
0.01s user 0.00s system 4% cpu 0.325 total
0.01s user 0.01s system 4% cpu 0.334 total
0.01s user 0.01s system 4% cpu 0.331 total
0.01s user 0.01s system 6% cpu 0.311 total
0.01s user 0.01s system 5% cpu 0.311 total
0.01s user 0.01s system 5% cpu 0.299 total
0.01s user 0.01s system 4% cpu 0.303 total
0.01s user 0.01s system 6% cpu 0.312 total
```

That doesn't seem too bad, especially since sharding could reduce this (not sure though, since there's the added overhead of additional requests).

What also comes on top is that `repo-updater` needs to load the names from the database

mrnugget commented:
On `k8s.sgdev.org` it takes ~6 seconds for one status message:

<img width="932" alt="Screen Shot 2019-06-19 at 14 12 52" src="https://user-images.githubusercontent.com/1185253/59764587-5deedf00-929c-11e9-8225-423a8e60277d.png">

Example trace: https://app.lightstep.com/sourcegraph-dev/trace?span_guid=39b1467c99d8683b&at_micros=1560946392318752#span-39b1467c99d8683b

I guess that could/should be faster.

The problem is though, that apparently 20 repositories are never cloned:

<img width="469" alt="Screen Shot 2019-06-19 at 14 13 21" src="https://user-images.githubusercontent.com/1185253/59764710-9ee6f380-929c-11e9-9833-318d0e25c242.png">

mrnugget commented:
The problem is that we're saying "Currently cloning" in the message, even though these repositories are only _enqueued_ for future cloning/updating. 

I could open the repository pages of 18 of the 20 repositories and they were cloned. The two others simply could not be cloned due to things like these:

```
error cloning repo" repo=github.com/gorilla/websocket err="error cloning repo: repo github.com/gorilla/websocket (ssh://git@phabricator.sgdev.org/diffusion/5/web.git) not cloneable: context deadline exceeded (output follows)\n\nssh: connect to host phabricator.sgdev.org port 22: Network unreachable\r\n
```

I then deleted the "broken" Phabricator external service now we're done to "all repositories up to date"

tsenart commented:
I agreed with @mrnugget to bring this PR to completion tomorrow since he's going to be off.

mrnugget commented:
I added a small in-memory cache for the "number of not-cloned repos" to repo-updater with a TTL of 30 seconds. 

From my side, this is ready to review and also ready to merge.

But **please note**: the behavior of the number changed slightly, as noted in the original PR description. When it now says "X repositories enqueued for cloning..." that doesn't mean that they are _currently_ cloning, but that they will be cloned at some point. Or in other words: the number matches the repositories in the repositories-page when the "not cloned" filter is activated, not the "Cloning" filter.

---

@christinaforney Note the textual changes in the diff. Since this now computes the actual number of repositories that have not been cloned yet, I changed the text to reflect that.

christinaforney commented:
@mrnugget - I like your wording change! I think it's important to be accurate in our descriptions as to what is going on, and this does that.

mrnugget commented:
@tsenart @keegancsmith Little thought I had waking up: we should probably bust this new cache when external services are synced.

---

And another thought I had that I want to investigate when I'm back: since we only want the number of not-cloned repos, how fast would it be to do a `SELECT COUNT(*) FROM repo;` on `repo-updater` and then either (a) use the existing `list?cloned` endpoint on the `gitserver` replicas (which walks through the directory to build a list of repos) or (b) use a new `number-of-cloned-repos` endpoint that also walks through the directory but only keeps count and doesn't return a huge list. My thinking is that walking through the file system might be slow, but we'd have a lot less pressure on the GC when we stop sending around lists of tens of thousands of repository names.

tsenart commented:
@mrnugget: Since you have a few ideas you'd like pursue further in this change set, and I have to focus on load testing Bitbucket Server ACLs today, would you feel comfortable picking this back up next week when you're back?

keegancsmith commented on cmd/repo-updater/repoupdater/server.go:
```diff
@@ -375,14 +384,21 @@ func (s *Server) shouldGetGithubDotComRepo(args protocol.RepoLookupArgs) bool {
 }
 
 func (s *Server) handleStatusMessages(w http.ResponseWriter, r *http.Request) {
+	if s.notClonedCountUpdatedAt.Before(time.Now().Add(-30 * time.Second)) {
```
need to protect this reads of this variable and notClonedCount with a mutex as well

keegancsmith commented on pkg/gitserver/client.go:
```diff
@@ -529,6 +529,76 @@ func (c *Client) IsRepoCloned(ctx context.Context, repo api.RepoName) (bool, err
 
+func (c *Client) AreReposCloned(ctx context.Context, repos ...api.RepoName) (*protocol.AreReposClonedResponse, error) {
+	numPossibleShards := len(c.Addrs(ctx))
+	shards := make(map[string]*protocol.AreReposClonedRequest, (len(repos)/numPossibleShards)*2) // 2x because it may not be a perfect division
```
size of shards is maximum `len(c.Addrs)`. Normally it will be len(c.Addrs) since we should have a atleast one repo per gitserver replica.

```suggestion
	shards := make(map[string]*protocol.AreReposClonedRequest, numPossibleShards)
```

keegancsmith commented:
> @tsenart @keegancsmith Little thought I had waking up: we should probably bust this new cache when external services are synced.

The cache expires in 30s. That should be short enough that its fine?

> And another thought I had that I want to investigate when I'm back: since we only want the number of not-cloned repos, how fast would it be to do a `SELECT COUNT(*) FROM repo;` on `repo-updater` and then either (a) use the existing `list?cloned` endpoint on the `gitserver` replicas (which walks through the directory to build a list of repos) or (b) use a new `number-of-cloned-repos` endpoint that also walks through the directory but only keeps count and doesn't return a huge list. My thinking is that walking through the file system might be slow, but we'd have a lot less pressure on the GC when we stop sending around lists of tens of thousands of repository names.

Yup that makes sense to me. But that could also be a follow-up PR since this works! :)

tsenart commented on cmd/repo-updater/repoupdater/server.go:
```diff
@@ -402,6 +417,37 @@ func (s *Server) handleStatusMessages(w http.ResponseWriter, r *http.Request) {
 }
 
+func (s *Server) computeNotClonedCount(ctx context.Context) (uint64, error) {
+	s.notClonedCountMu.Lock()
```
This lock is too coarse. Can we lock only around the critical sections and avoid locking around slow RPCs?

keegancsmith replied:
by doing it around the slow RPCs we get single flighting for free.

tsenart replied:
It wasn't obvious that you wanted single flighting. Can you add a comment explaining why that's desirable here?

tsenart commented on cmd/repo-updater/repos/store.go:
```diff
@@ -26,6 +26,8 @@ type Store interface {
 	ListRepos(context.Context, StoreListReposArgs) ([]*Repo, error)
 	UpsertRepos(ctx context.Context, repos ...*Repo) error
+
+	ListAllRepoNames(context.Context) ([]api.RepoName, error)
```
I'm fine with adding this extra method if it's necessary for the performance gains we need.

mrnugget commented:
Thanks for the changes @keegancsmith and bringing this over the finishing line :)
